packages/salesforcedx-vscode-core/src/conflict/timestampConflictDetector.ts

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { join, relative } from 'path';
+import { nls } from '../messages';
+import {
+  DirectoryDiffResults,
+  MetadataCacheResult,
+  MetadataCacheService,
+  PersistentStorageService
+} from './';
+import { ComponentDiffer } from './componentDiffer';
+import { CorrelatedComponent } from './metadataCacheService';
+
+export class TimestampConflictDetector {
+  private differ: ComponentDiffer;
+  private diffs: DirectoryDiffResults;
+  // Probably want a UI update to remove scannedLocal and scannedRemote
+  private static EMPTY_DIFFS = {
+    localRoot: '',
+    remoteRoot: '',
+    different: new Set<string>(),
+    scannedLocal: 0,
+    scannedRemote: 0
+  };
+
+  constructor() {
+    this.differ = new ComponentDiffer();
+    this.diffs = Object.assign({}, TimestampConflictDetector.EMPTY_DIFFS);
+  }
+
+  public createDiffs(
+    result?: MetadataCacheResult
+  ): DirectoryDiffResults {
+    if (!result) {
+      throw new Error(nls.localize('conflict_detect_empty_results'));
+    }
+    this.createRootPaths(result);
+    const components = MetadataCacheService.correlateResults(result);
+    this.determineConflicts(components);
+    return this.diffs;
+  }
+
+  private determineConflicts(
+    data: CorrelatedComponent[]
+  ) {
+    const cache = PersistentStorageService.getInstance();
+    const conflicts: Set<string> = new Set<string>();
+    data.forEach(component => {
+      let lastModifiedInOrg;
+      let lastModifiedInCache;
+
+      lastModifiedInOrg = component.fileProperties.lastModifiedDate;
```

Comment:
Help me understand this - why are we using fileProperties here? Didn't we discuss about using SourceComponents for both retrieve and deploy. The same question for the "correlateResults()" method - why we are retrieving and storing fileProperties?

Reply:
Good question! We need FileProperties for thing, and that is the lastModifiedDate. For the cache keys, we decided to start using type name and fullName because those are fields that exist on both FileProperties and SourceComponents and sometimes, such as after a deploy, we don't have FileProperties. You'll see that in this PR, FileProperties were added to the MetadataCacheResult just so we can get access to the lastModifiedDate here. 

A valid change we could make is only storing lastModifiedDate instead of the whole FileProperties in the CorrelatedComponent type.

Reply:
I like the idea of storing lastModifiedDate instead of the whole FileProperties in the CorrelatedComponent type.
One more question - where do we get lastModifiedDate after deploy given that FileProperties is not available?

Reply:
It is just a field in the MetadataApiDeployStatus (the supposed reason being that FileProperties are not necessary, they would be the exact same for every file getting deployed).

Reply:
Made this change in a recent [commit](https://github.com/forcedotcom/salesforcedx-vscode/pull/3332/commits/c2fe3428dc4c1eb7654b5e42a13f70c0dd132ada)